evm/src/cpu/shift.rs

```diff
@@ -0,0 +1,139 @@
+use plonky2::field::extension::Extendable;
+use plonky2::field::packed::PackedField;
+use plonky2::field::types::Field;
+use plonky2::hash::hash_types::RichField;
+use plonky2::iop::ext_target::ExtensionTarget;
+
+use crate::constraint_consumer::{ConstraintConsumer, RecursiveConstraintConsumer};
+use crate::cpu::columns::CpuColumnsView;
+use crate::cpu::membus::NUM_GP_CHANNELS;
+use crate::memory::segments::Segment;
+
+pub(crate) fn eval_packed<P: PackedField>(
+    lv: &CpuColumnsView<P>,
+    yield_constr: &mut ConstraintConsumer<P>,
+) {
+    let is_shift = lv.op.shl + lv.op.shr;
+    let val = lv.mem_channels[0];
+    let displacement = lv.mem_channels[1]; // holds the shift displacement d
+    let two_exp = lv.mem_channels[2]; // holds 2^d
+    let output = lv.mem_channels[NUM_GP_CHANNELS - 1]; // should hold val * 2^d (mod 2^256)
+
+    let shift_table_segment = P::Scalar::from_canonical_u64(Segment::ShiftTable as u64);
+
+    // Value, displacement, 2^disp, and output channels must be used and read-only.
+    yield_constr.constraint(is_shift * (val.used - P::ONES));
+    yield_constr.constraint(is_shift * (val.is_read - P::ONES));
+    yield_constr.constraint(is_shift * (displacement.used - P::ONES));
+    yield_constr.constraint(is_shift * (displacement.is_read - P::ONES));
+    yield_constr.constraint(is_shift * (output.used - P::ONES));
+    yield_constr.constraint(is_shift * (output.is_read - P::ONES));
+    yield_constr.constraint(is_shift * (two_exp.used - P::ONES));
+    yield_constr.constraint(is_shift * (two_exp.is_read - P::ONES));
+
+    let high_limbs_are_zero = lv.general.shift().displacement_high_limbs_are_zero;
+    yield_constr
+        .constraint(is_shift * (high_limbs_are_zero - high_limbs_are_zero * high_limbs_are_zero));
+
+    let high_limbs_sum: P = displacement.value[1..].iter().copied().sum();
+    yield_constr.constraint(is_shift * high_limbs_sum * high_limbs_are_zero);
+
+    // When the shift displacement is < 2^32, constrain the two_exp
+    // mem_channel to be the entry corresponding to `displacement` in
+    // the shift table lookup (will be zero if displacement >= 256).
+    let small_disp_filter = is_shift * high_limbs_are_zero;
+    yield_constr.constraint(small_disp_filter * two_exp.addr_context); // kernel mode only
+    yield_constr.constraint(small_disp_filter * (two_exp.addr_segment - shift_table_segment));
+    yield_constr.constraint(small_disp_filter * (two_exp.addr_virtual - displacement.value[0]));
+
+    // When the shift displacement is >= 2^32, constrain two_exp.value
+    // to be zero.
+    let large_disp_filter = is_shift * (P::ONES - high_limbs_are_zero);
+    for &limb in &two_exp.value {
+        yield_constr.constraint(large_disp_filter * limb);
+    }
```

Comment:
This doesn’t quite work. The `two_exp` memory channel is always used, so if you constrain its `value` to zero, then it might not match the contents of memory (so you can’t satisfy both constraints).

You can fix it by making `two_exp.used` iff `high_limbs_are_zero`:
```rust
yield_constr.constraint(is_shift * (two_exp.used - high_limbs_are_zero))
```

Reply:
Oops, good catch!